sys/kern/kftrace.c

```diff
@@ -0,0 +1,99 @@
+#include <sys/kftrace.h>
+#include <sys/thread.h>
+#include <sys/types.h>
+#include <sys/malloc.h>
+
+/* Kernel Function Trace
+ *
+ * In this file there are defined structures and functions that are used to
+ * implement function tracking in kernel.
+ *
+ * When entering or exiting a function an event is added to `kft_event_list`.
+ * The size of the list is specified by `KFT_EVENT_MAX`.
+ *
+ * When the list is full the function `kft_flush` is invoked. Currently it only
+ * resets the event count. The main purpose of it is to set a breakpoint on it
+ * and dump the contents of the list of events to a file by the debugger.
+ *
+ * Information about the event is compressed into a single 64 bit value.
+ * Below is a picture of layout of information encoded in that value.
+ *
+ * +---------+-----------+-----------+------------+
+ * | 24 bits | 31 bits   |   8 bits  |   1 bit    |
+ * |    PC   | timestamp | thread id | event type |
+ * +---------+-----------+-----------+------------+
+ */
+
+typedef uint64_t kft_event_t;
+
+#ifdef MIPS_KFT
+#define KFT_EVENT_MAX 8000
+#else
+#define KFT_EVENT_MAX 1000000
+#endif
+
+static kft_event_t *kft_event_list;
+static unsigned kft_used = 0;
+static bool kft_enabled = false;
+
+#define PC_MASK 0xFFFFFF          /* 24 bits */
```

Comment:
Please note that lower two bits won't be used on all architectures with 4-byte only instruction words (i.e. all that we support currently).